confdgnmi/tests/test_client_server.py

```diff
@@ -166,22 +248,175 @@ def verify_sub_updates(paths, pv_list,
+                sleep(1)
+            else:
+                path_prefix = make_gnmi_path(prefix_str)
+                path = make_gnmi_path(c[0])
+                cmd = "{} {}".format(
+                    make_formatted_path(path, gnmi_prefix=path_prefix),
+                    c[1])
+                if "state" in prefix_str:
+                    oper_cmd += "{} ".format(cmd)
+                else:
+                    config_cmd += "mset {};".format(cmd)
+        log.info("<==")
+
+    @pytest.mark.long
+    @pytest.mark.parametrize("adapter_type",
+                             [pytest.param(AdapterType.DEMO,
+                                           marks=[pytest.mark.demo]),
+                              pytest.param(AdapterType.API,
+                                           marks=[pytest.mark.confd])
+                              ])
+    @pytest.mark.parametrize("data_type", ["CONFIG", "STATE"])
+    def test_subscribe_stream(self, request, adapter_type, data_type):
+        log.info("testing subscribe_stream")
+
+        if_state_str = prefix_state_str = ""
+        if data_type == "STATE":
+            prefix_state_str = "-state"
+            if_state_str = "state_"
+        prefix_str = "/interfaces{}".format(prefix_state_str)
+        prefix = make_gnmi_path(prefix_str)
+        paths = [TestGrpc.mk_gnmi_if_path(self.list_paths_str[1], if_state_str,
+                                          "N/A")]
+        changes_list = [
+            ("interface[name={}if_5]/type".format(if_state_str),
+             "fastEther"),
+            ("interface[name={}if_6]/type".format(if_state_str),
+             "fastEther"),
+            "send",
+            ("interface[name={}if_5]/type".format(if_state_str),
+             "gigabitEthernet"),
+            ("interface[name={}if_6]/type".format(if_state_str),
+             "gigabitEthernet"),
+            "send",
+        ]
+        if data_type == "STATE" and adapter_type == AdapterType.API:
+            # TODO state `confd_cmd` is not transactional, so we need to check
+            # every item - add 'send' after each item (fix checking method?)
+            new_change_list = []
+            must_send = True
+            for c in changes_list:
+                if c == "send":
+                    must_send = False
+                new_change_list.append(c)
+                if must_send:
+                    new_change_list.append("send")
+                    must_send = False
+                else:
+                    must_send = True
+            changes_list = new_change_list
+        log.info("change_list=%s", changes_list)
+
+        # create  config string
+        if adapter_type == AdapterType.DEMO:
+            demo = ET.Element("demo")
+            sub = ET.SubElement(demo, "subscription")
+            stream = ET.SubElement(sub, "STREAM")
+            changes = ET.SubElement(stream, "changes")
+
+        path_value = [[]]  # empty means no check
```

Comment:
why adding "send" elements before, and now skipping them again?
create non-send list of changes only, do what's necessary, and add "sends" in final stage?

Reply:
sends are added only for "STATE" type and API adapter.
I would like to have structure similar to XML config file, with support for more elements later on.